[TG-2608] Use string dependencies to reduce the number of constraints

romainbrenguier opened the pull request:
Instead of adding all constraints corresponding to a string function
call, we look at which strings may be used in a testing function (such
as endsWith, equals, compare, etc...). 
We only add constraints for these function calls, since the other operations can not matter for the
satisfiability problem.

romainbrenguier committed:
Add class for builtin function with no eval

This is for builtin functions supported by string_constraint_generatort
but not string_builtin_functiont.

romainbrenguier committed:
Constructor for builtin function with no eval

romainbrenguier committed:
never return nullptr in to_string_builtin_function

The default is now to return a function with no eval.
This also change add a use of util_make_unique in the insert case.

romainbrenguier committed:
Remove use of unknown_dependency

unknown dependencies on a builtin function are now using
string_builtin_function_with_no_evalt.

romainbrenguier committed:
Add and add_constraint method to builtin functions

This will be used to directly add constraints by traversing the string
dependency graph.

romainbrenguier committed:
add_constraints method in string_dependencies

romainbrenguier committed:
Add constraints from the string dependencies

We add constraints directly from the constructed graph instead of going
through all the equations.

romainbrenguier committed:
Initialize return_code for all builtin_functions

romainbrenguier committed:
Store builtin function pointer inside builtin node

This is more natural than always having to refer to the
builtin_function_nodes array.

romainbrenguier committed:
Add a maybe_testing_function to builtin functions

This is to know whether the result of a function can be used outside of
string operations.

romainbrenguier committed:
Only add constraints on test function dependencies

Instead of adding all constraints corresponding to a string function
call, we look at which strings may be used in a testing function (such
as endsWith, equals, compare, etc...). We only add constraints for these
function calls, since the other operations cannot matter for the
satisfiability problem.

romainbrenguier committed:
Generic get_reachable function

This takes as argument a set and a for_each_successor function.
This can be used even by classes that do not inherit from grapht.

romainbrenguier committed:
Add a hash function for nodes

This is to be able to use containers such as unordered_set/unordered_map

romainbrenguier committed:
Add tests for the use of string dependencies

This tests, check that only constraints for the string operation that
matters are added to the list of axioms of the string solver.
In the example added here, the solver would previously consider 11
constraints, but now only 3, and still finds the correct result.

tautschnig commented on src/util/graph.h:
```diff
@@ -458,6 +458,34 @@ void grapht<N>::visit_reachable(node_indext src)
+/// \param set: set of source nodes, must be a container with an
+///   `insert(const value_type&)` method.
+/// \param for_each_successor: function which given a node `n` and a function
+///   `f`, applies `f` on all successors of `n`.
```
Could this comment please be a bit more technical and actually say which graph traversal algorithm is used (I think it's depth-first search), and explain what the idea behind this function is?

romainbrenguier replied:
I updated the comment

tautschnig commented on src/util/graph.h:
```diff
@@ -458,6 +458,34 @@ void grapht<N>::visit_reachable(node_indext src)
+    const std::function<void(const typename Container::value_type &)> &)>
+    &for_each_successor)
+{
+  std::vector<nodet> stack;
```
Don't store `nodet`, use `node_indext`. Going from a `node_indext` to a `nodet` is constant time, but a `node_indext` will typically consume a lot less memory.

romainbrenguier replied:
Since this function is generic you can use it different ways, in particular if you give a set of node_indext then the algorithm will only manipulate node_indext. nodet here does not have to correspond to graph::node_t.

tautschnig replied:
Got it.

tautschnig commented on src/util/graph.h:
```diff
@@ -458,6 +458,34 @@ void grapht<N>::visit_reachable(node_indext src)
+/// \param for_each_successor: function which given a node `n` and a function
+///   `f`, applies `f` on all successors of `n`.
+template <class Container, typename nodet = typename Container::value_type>
+void get_reachable(
```
~This function is introduced, but never used in this PR?~

romainbrenguier replied:
It used in https://github.com/diffblue/cbmc/pull/1947/commits/b2ca93723b89c236b8ceb5be3caa2266e60bb998

tautschnig reviewed (commented):
I have only reviewed the changes to graph.h.

romainbrenguier committed:
Add move constructors to builtin_function_nodet

Because of some issues with Visual Studio these need to be defined.

romainbrenguier committed:
Rewrite axioms for insert

This reduces the number of universal formulas added (from 6 to 3), and
avoid introducing new strings. Ideally the add_axioms_-function should
not introduce new strings, that would make it easier to refactor and add
new functionalities.

romainbrenguier committed:
Correct name of builtin function with no eval

romainbrenguier commented:
I extracted a few commits which correct existing bugs in https://github.com/diffblue/cbmc/pull/1958 which should be merged before this PR.

romainbrenguier committed:
Adapt unit tests for the new interface

The interface for string dependencies has changed so these need to be adapted.

romainbrenguier committed:
Define two helper functions for for_each_successor

These two functions deal with one type of node.
Type safety ensures we don't mix the two kinds of node, so these two
functions can be made public.

romainbrenguier committed:
Add double quotes to output_dot

LAJW commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -336,44 +336,57 @@ bool add_node(
+  std::for_each(
+    node.dependencies.begin(),
+    node.dependencies.end(),
+    [&](const std::size_t &index) { f(builtin_function_nodes[index]); });
```
It looks like you could just use a ranged for-loop.

LAJW commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -336,44 +336,57 @@ bool add_node(
   else
-    UNREACHABLE;
+  {
+    INVARIANT(node.kind == nodet::STRING, "nodes are either BUILTIN or STRING");
```
I think UNREACHABLE at the end would be better.

romainbrenguier committed:
Deactivating invariant

peterschrammel commented on src/solvers/refinement/string_builtin_function.h:
```diff
@@ -32,8 +33,29 @@ class string_builtin_functiont
+  /// Tells whether the builtin function can be a testing function, that is a
+  /// function that does not return a string, for instance like `equals`,
+  /// `indexOf` or `compare`.
+  virtual bool maybe_testing_function() const
```
Why `maybe` and not `is`?

romainbrenguier replied:
For the functions that don't have yet a dedicated builtin_function class we consider that could be testing function although we don't actually know. Ultimately this can be replaced by a `is`

peterschrammel commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -387,42 +336,59 @@ bool add_node(
+        f(*string_node);
+      else
+      {
+        std::cout << "Warning no node for " << format(current) << std::endl;
```
`cout` is forbidden outside `#ifdef DEBUG`

peterschrammel commented on src/solvers/refinement/string_refinement_util.cpp:
```diff
@@ -387,42 +336,59 @@ bool add_node(
+      else
+      {
+        std::cout << "Warning no node for " << format(current) << std::endl;
+        //UNREACHABLE;
```
Why is this commented out?

romainbrenguier replied:
I reestablish it in a later commit that could be squashed with this one.
 https://github.com/diffblue/cbmc/pull/1947/commits/ec6321f7e2a0ffeab0d4071f4a2b9b94543af604

romainbrenguier committed:
Helper functions for length constraints on strings

This defines functions adding constraints on length of strings.
This constraints are used in the add_axioms_* functions and the goal is
to reuse them in string dependencies, to add axioms only about the
length and not the content of the strings.

romainbrenguier committed:
Add length_constraint method to builtin functions

This give the constraints that ensure the length of the result of a
buitin function is correct. Will be used when we don't need the
constraints about the actual content of the strings.

romainbrenguier committed:
Correct add_constraints in string_dependencies

When the result of a function is not tested we don't need to add full
constraints, but constraints about the length may still be necessary as
the length field of the string can be accessed directly.

romainbrenguier committed:
Define a for_each_atomic_string helper function

This could be used in several places in this file.

romainbrenguier committed:
Ensure all atomic strings in arguments have node

We make sure a node is created for each atomic string in arguments of
builtin functions. This allows us to restablish the check that a node
exists for each dependency of a builtin function.

peterschrammel reviewed (commented):
There's a failing test: https://ci.appveyor.com/project/diffblue/cbmc/build/1.0.8107#L5008

romainbrenguier committed:
Clear constraints at each call to dec_solve

constraint_generator is kept as a class member because it is used by
get, but we need to make sure it is cleared at each dec_solve call so
that we don't keep unecessary or redundant constraints.

romainbrenguier committed:
Reduce string-max-length on String.equals test

Execution time can vary depending on what the sat solver find as a
model at the refinement steps.
Observed execution time go from 0.2s to 7s.
Setting the limit to 100 limit these variations (maximum observed is
0.3s).
